### DISCUSSION TITLE
Reduce the time to simulate grain growth

### DISCUSSION THREAD

**Discussion by PengWei97** [idaholab/moose#16592] [Q&A Modules: Phase field] [answered]
Dear MOOSE experts,
I am trying to simulate a grain growth simulation of 1000 grains. I set end_time = 10000, but it took 58 hours. How can I optimize the input file to reduce the running time?
The following are the basic parameters and related files used in my simulation:
Number of cpu cores: 20
Number of initial grids: 409600
Grid adaptation and time adaptation are used
Input file: [grain_growth_2D_graintracker_1000_output.txt](https://github.com/idaholab/moose/files/5742670/grain_growth_2D_graintracker_1000_output.txt)
Part of the information output to the terminal: 
[grain_growth_2D_graintracker_1000_output.log](https://github.com/idaholab/moose/files/5742671/grain_growth_2D_graintracker_1000_output.log)
After adapting the grid, the initial model diagram：
![Snipaste_2020-12-26_12-47-08](https://user-images.githubusercontent.com/56183775/103145684-155d8f00-4779-11eb-9a36-f8d0b177d51c.png)

Any suggestions or recommendations to fix these problems would be greatly appreciated.
Thank you
Wei Peng

**Comment by laagesen** [comment]
A couple of days is not unreasonable for a 1000-grain simulation. There are a couple of things you could try:
-increase `optimal_iterations` to 8 or 10 in the TimeStepper block. This will likely give you larger time steps. You should check to make sure the results are consistent with the initial run you did with that parameter set to 6
-Add a NumDOFs postprocessor so you know how many DOFs are in your problem as a function of time. You generally want about 10,000 to 20,000 DOF per core, so you can choose your # of cores to try to maintain this. When you say `Number of initial grids: 409600` I am not sure if you mean number of elements in the mesh or number of DOFs. If it is # of elements in the mesh, the number of DOFs is probably significantly larger, so you are probably much larger that 20,000 DOF/core since you are using 20 cores. So you would need more cores.

**Reply by PengWei97** [reply]
Hi laagesen,
Thank you very much for your response. Yes, `Number of initial grids: 409,600` means number of elements. And `the number of DOFs: 4,108,810`, which is displayed in the log file at the end. According to your suggestion, the number of cores used is about 205, which is a very large number! 

The question I want to ask is, is 205 the optimal number of cores to run this example?

Thanks again for your time. Finally, do you have some suggestions to optimize the input file to shorten the simulation time for a beginner like me? For example, whether to use a distributed mesh or a replicated mesh, etc.

Kind regards,
Wei Peng

Details are as follows,

```
Framework Information:
MOOSE Version:           git commit 54c00d0775 on 2020-12-17
LibMesh Version:         c176405e538983b35afc4f6827e29e155141e42e
PETSc Version:           3.14.2
SLEPc Version:           3.14.0
Current Time:            Wed Dec 23 22:33:21 2020
Executable Timestamp:    Sun Dec 20 22:02:49 2020

Parallelism:
  Num Processors:          10
  Num Threads:             1

Mesh: 
  Parallel Type:           replicated
  Mesh Dimension:          2
  Spatial Dimension:       2
  Nodes:                   
    Total:                 410881
    Local:                 41401
  Elems:                   
    Total:                 409600
    Local:                 40973
  Num Subdomains:          1
  Num Partitions:          10
  Partitioner:             metis

Nonlinear System:
  Num DOFs:                4108810
  Num Local DOFs:          414010
  Variables:               { "gr0" "gr1" "gr2" "gr3" "gr4" "gr5" "gr6" "gr7" "disp_x" "disp_y" } 
  Finite Element Types:    "LAGRANGE" 
  Approximation Orders:    "FIRST" 

Auxiliary System:
  Num DOFs:                3687681
  Num Local DOFs:          369185
  Variables:               "bnds" { "elastic_strain11" "elastic_strain22" "elastic_strain12" "unique_grains" 
                             "var_indices" "vonmises_stress" "C1111" "euler_angle" } 
  Finite Element Types:    "LAGRANGE" "MONOMIAL" 
  Approximation Orders:    "FIRST" "CONSTANT" 

Execution Information:
  Executioner:             Transient
  TimeStepper:             IterationAdaptiveDT
  Solver Mode:             Preconditioned JFNK
  MOOSE Preconditioner:    SMP
```

**Comment by friedmud** [comment, marked as answer]
I agree with @laagesen : the best way to speed this up is to use more processors.  As he mentioned, you could easily use 200 procs with 4M dofs.

As for "optimal" - that really depends.  We recommend about 20k dofs/proc in order to keep parallel efficiency high.  You may be able to go well beyond that and still get speedup (I would expect to see speedup on this problem out to 500 cores or so)... but your parallel efficiency will start to drop off (the amount of speed increase won't be commensurate with the number of procs you're using).

Another thing you may want to try is to use the DT2 timestepper... you will have to play with the parameters to see what works for your problem - but there is a possibility of greatly increasing the timestep size.

**Reply by PengWei97** [reply]
Thank friedmud for your information. Regarding the second suggestion you gave, I don’t quite understand it. I don't quite understand your second suggestion. You mean to add the TimeStepper module to increase the size of the time step, right?

Finally, if I change the Mesh type to a [distributed mesh](https://mooseframework.inl.gov/syntax/Mesh/index.html), will this help reduce simulation time? And, in MOOSE, which aspects of the calculation are affected by the mesh type?

**Reply by GiudGiud** [reply]
The DT2 time stepper is an adaptive time stepper that will increase the time step through the simulation, based on comparing the solution with time steps of dt and dt/2. This is usually more effective than trying to uniformly increase the time step. 

Using a distributed mesh is typically done to reduce the memory footprint, as it avoids duplicating the mesh for every process. I don't foresee a large impact on simulation time. 
EDIT: Maybe with the mesh adaptivity, there could be an impact

You could try running your simulation with the --timing flag (EDIT add PerfGraphOutput) to see which parts of the simulation are the most computationally expensive.

**Reply by PengWei97** [reply]
After adding PerfGraphOutput to the output, it displays the proportion of each calculation part on the terminal as follows,

![image](https://user-images.githubusercontent.com/56183775/103890659-7004cc80-5123-11eb-9aea-f3fc63a41b25.png)

![image](https://user-images.githubusercontent.com/56183775/103890691-80b54280-5123-11eb-897a-4d35114009a2.png)

After adding PerfGraphOutput to the output, it displays the proportion of each calculation part on the terminal as follows,
In this regard, would you have any good suggestions for moose beginners like me?

**Reply by GiudGiud** [reply]
The previous advice given by @laagesen and @friedmud still applies. 

In addition, it seems to spend a significant amount of time in mesh adaptivity. Maybe using a distributed mesh is a good idea? I am not sure how the modifications to the mesh are spread across processes. 

Since the problem is converging with no issues, you could use the skip_exception_check parameter in your executioner to gain 5% runtime here. 
EDIT: It's probably not really that function causing the 5% overhead, it's probably just work imbalance between nodes that's felt at this synchronization step

**Reply by PengWei97** [reply]
Thank you for your suggestion. I will try to use a distributed mesh and hope it will help reduce the calculation time.
